File: classes/c++ files/C++/CardPayment.cpp
#include "CardPayment.h"
#include <iostream>
#include <iomanip>

CardPayment::CardPayment(float amount, string last4Digits, string card) : Payment(amount) {
    last4    = last4Digits;
    cardType = card;
}

CardPayment::~CardPayment() {}

// Getters
string CardPayment::getLast4()    { return last4; }
string CardPayment::getCardType() { return cardType; }

// Setters
void CardPayment::setLast4(string last4)       { this->last4    = last4; }
void CardPayment::setCardType(string cardType) { this->cardType = cardType; }

bool CardPayment::processPayment() {
    cout << "Card approved." << endl;
    return true;
}

void CardPayment::displayPaymentInfo() {
    cout << fixed << setprecision(2);
    cout << "Card Type:    " << cardType         << endl;
    cout << "Card Number:  **** **** **** " << last4 << endl;
    cout << "Amount:       $" << getAmount()     << endl;
}

File: classes/CardPayment.ts
import { PaymentDef } from "./PaymentDef";

export class CardPayment extends PaymentDef {
  private last4: string;
  private cardType: string;

  constructor(amount: number, last4: string, cardType: string) {
    super();
    this.setAmount(amount);
    this.last4    = last4;
    this.cardType = cardType;
  }

  getAmount(): number             { return this.amount; }
  setAmount(amount: number): void { this.amount   = amount; }
  getLast4(): string              { return this.last4; }
  getCardType(): string           { return this.cardType; }
  setLast4(last4: string): void       { this.last4    = last4; }
  setCardType(cardType: string): void { this.cardType = cardType; }

  processPayment(): boolean {
    return true; // simulated approval
  }

  displayPaymentInfo(): void {
    console.log(`Card Type:   ${this.cardType}`);
    console.log(`Card Number: **** **** **** ${this.last4}`);
    console.log(`Amount:      $${this.amount.toFixed(2)}`);
  }
}
